hpx/lcos/future.hpp

```diff
@@ -1059,35 +961,16 @@ namespace hpx { namespace lcos
                 std::forward<F>(f), ec);
         }
 
-        template <typename Policy, typename F>
+        template <typename T0, typename F>
         typename util::lazy_enable_if<
-            hpx::traits::is_launch_policy<
-                typename std::decay<Policy>::type
-            >::value,
+            !hpx::traits::is_invocable<F(future&&)>::value,
             hpx::traits::future_then_result<future, F>
         >::type
-        then(Policy && policy, F && f, error_code& ec = throws)
+        then(T0 && t0, F && f, error_code& ec = throws)
         {
             invalidate on_exit(*this);
             return base_type::then(std::move(*this),
-                std::forward<Policy>(policy), std::forward<F>(f), ec);
-        }
-
-        template <typename Executor, typename F>
-        typename util::lazy_enable_if<
-            hpx::traits::is_one_way_executor<
-                typename std::decay<Executor>::type>::value ||
-            hpx::traits::is_two_way_executor<
-                typename std::decay<Executor>::type>::value ||
-            hpx::traits::is_threads_executor<
-                typename std::decay<Executor>::type>::value
-          , hpx::traits::future_then_executor_result<Executor, future, F>
```

Comment:
@hkaiser `future_then_executor_result` remains as an unwanted dependency. Would you suggest we:
1. Use `future_then_result` instead and get rid of `future_then_executor_result`? Are there actually cases where `async_execute` would return something other than `future_then_result`, and is this something we want to keep allowing?
2. We introduce a `future_then_T0_result` where `T0` is some useful name describing something policy- or executor-like, which we then specialize in the executor module again like for `future_then_dispatch`?

The second sounds like it's definitely doable, but I'd rather simplify *if* we don't actually to specialize the return type per executor.

Reply:
It looks like using `future_then_result` for the launch policy/executor overload does indeed work in our codebase. The question is then if that'll always be correct and if we still want to keep `future_then_T0_result` in some form.

Reply:
@hkaiser any comments on this ?

Reply:
Ping @hkaiser. As this works I'd like to move on with this PR. We can decide on what trait to use here later as well.